presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomShuffleProbeSide.java

```diff
@@ -0,0 +1,173 @@
+
+    public RandomShuffleProbeSide(Metadata metadata, FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+    }
+
+    @Override
+    public void setEnabledForTesting(boolean isSet)
+    {
+        isEnabledForTesting = isSet;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isEnabledForTesting || !getRandomizeProbeSideStrategy(session).equals(DISABLED);
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        if (isEnabled(session)) {
+            return SimplePlanRewriter.rewriteWith(new Rewriter(session, metadata, functionAndTypeManager, idAllocator, variableAllocator), plan, new HashSet<>());
+        }
+
+        return plan;
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Set<VariableReferenceExpression>>
+    {
+        private final Session session;
+        private final Metadata metadata;
+        private final FunctionAndTypeManager functionAndTypeManager;
+        private final PlanNodeIdAllocator planNodeIdAllocator;
+        private final VariableAllocator planVariableAllocator;
+        private final FeaturesConfig.RandomizeProbeSideStrategy strategy;
+
+        private Rewriter(Session session, Metadata metadata,
+                FunctionAndTypeManager functionAndTypeManager, PlanNodeIdAllocator planNodeIdAllocator, VariableAllocator planVariableAllocator)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+            this.planNodeIdAllocator = requireNonNull(planNodeIdAllocator, "planNodeIdAllocator is null");
+            this.planVariableAllocator = requireNonNull(planVariableAllocator, "planVariableAllocator is null");
+            this.strategy = getRandomizeProbeSideStrategy(session);
+        }
+
+        private static boolean isBroadcastJoin(JoinNode joinNode)
+        {
+            return joinNode.getDistributionType().isPresent() && joinNode.getDistributionType().get() == REPLICATED;
+        }
+
+        private static boolean isApplicable(JoinNode joinNode, FeaturesConfig.RandomizeProbeSideStrategy strategy)
+        {
+            return isBroadcastJoin(joinNode) &&
+                    (strategy == FeaturesConfig.RandomizeProbeSideStrategy.ALWAYS ||
+                            strategy == FeaturesConfig.RandomizeProbeSideStrategy.COST_BASED && isLeftSideSkewed(joinNode));
+        }
+
+        private static boolean isLeftSideSkewed(JoinNode joinNode)
```

Comment:
Curious and just trying to learn so a n00b q: will Hbo also be giving us what keys are skewed ?

Reply:
we don't plan to store skewed key values at the moment due to privacy implications regarding user data; rather we will store flags whether the data is skewed or not